### PR TITLE
[FLINK-5763] [checkpoints] Add CheckpointOptions for self-contained savepoints

### PR DESCRIPTION
**Pull request by uce**
This is based on #3345 and only the last two commits are relevant here. I've separated the test changes (last commit) and the main changes (2nd last commit) for better reviewability. I would squash them before merging this. The change is looks more involved than it actually is. It is mostly routing new information with the checkpointing barriers, which touches a lot of places.

The main change is to add `CheckpointOptions` to the triggered checkpoint messages (coordinator to barrier injecting tasks) and barriers (flowing inline with the data):

```java
public class CheckpointOptions {
	
	// Type of checkpoint
	// => FULL_CHECKPOINT
	// => SAVEPOINT
	@NonNull
	CheckpointType getCheckpointType();

	// Custom target location. This is a String, because for future
	// backends it can be a logical location like a DB table.
	@Nullable
	String getTargetLocation();

}
```

This class would be the place to define more options for performing the checkpoints (for example for incremental checkpoints). @StephanEwen was involved with the design of incremental checkpoints and could probably comment best whether this is inline with the design for that.

These options are forwarded via the `StreamTask` to the `StreamOperator`s and `Snapshotable` backends. The `AbstractStreamOperator` checks the options and either
i) forwards the shared per operator `CheckpointStreamFactory` (as of #3312), or
ii) creates a custom savepoint stream factory (one per savepoint).

For this, the state backends provide the following new method:

```java
CheckpointStreamFactory createSavepointStreamFactory(JobID, String, String);
```

The `MemoryStateBackend` returns the regular stream factory and the `FsStateBackend` returns a `FsSavepointStreamFactory`, which writes all checkpoint streams to a single directory (instead of the regular sub folders per checkpoint).

We end up with the following directory layout for savepoints:

```
+---------------------------+
| :root_savepoint_directory | (custom per savepoint or configured default via `state.savepoints.dir`)
+---------------------------+
  | +---------------------------------------+
  +-| savepoint-:jobId(0, 6)-:random_suffix | (one directory per savepoint)
    +---------------------------------------+
      |
      +- _metadata (one per savepoint)
      +- :uuid (one data file per StreamTask)
      +- ...
      +- :uuid
```

I decided to include a prefix of the job ID to the savepoint directory, because I think that this could be helpful to map savepoints to jobs, which is a manual task. 

It's important to make sure that this is inline with upcoming changes for incremental checkpoints (discussion on mailing list) and FLINK-5820.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 1]
```diff
@@ -296,15 +298,42 @@ public boolean isShutdown() {
...
-		CheckpointTriggerResult result = triggerCheckpoint(timestamp, props, targetDirectory, false);
 
-		if (result.isSuccess()) {
-			return result.getPendingCheckpoint().getCompletionFuture();
-		}
-		else {
-			Throwable cause = new Exception("Failed to trigger savepoint: " + result.getFailureReason().message());
-			return FlinkCompletableFuture.completedExceptionally(cause);
+		// Create the unique savepoint directory
+		final String savepointDirectory = SavepointStore
```
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java]: We could also not create the directory here, but instead simply create the path name and send it to the tasks.

@StephanEwen Maybe creating directories like these is something we should keep in mind for the state backends on the JobManager?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 2]
```diff
@@ -431,11 +436,12 @@ public void snapshotState(StateSnapshotContext context) throws Exception {
 	@SuppressWarnings("deprecation")
 	@Deprecated
 	@Override
-	public StreamStateHandle snapshotLegacyOperatorState(long checkpointId, long timestamp) throws Exception {
+	public StreamStateHandle snapshotLegacyOperatorState(long checkpointId, long timestamp, CheckpointOptions checkpointOptions) throws Exception {
 		if (this instanceof StreamCheckpointedOperator) {
+			CheckpointStreamFactory factory = getCheckpointStreamFactory(checkpointOptions);
```
Comment (uce) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: Here we create the factory again for legacy state. As an alternative, we would have to create the factory in the `StreamTask` and pass it to the snapshot call here.

**Comment by gyfora** [comment]
So will the same savepoint logic apply to externalized checkpoints? I think it would be good to have a similar way of restoring from checkpoints and savepoints from a usability perspective. 

I usually set the externalized checkpoint dir the same as the savepoint dir, to make it easy to write scripts to get the latest and restart.

Otherwise I think the changes make a lot of sense and the directory structure looks very reasonable. 

One question, in the directory name could we maybe use the checkpoint id instead of the random suffix or something more predictable? Maybe the checkpoint date?

**Comment by StephanEwen** [comment]
This is a prerequisite for https://issues.apache.org/jira/browse/FLINK-5820
Reviewing this now...

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java (hunk 1 above)
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java]: Let's keep it as it is for this PR, will probably change a bit once the for the JobManager knows the StateBackend,

**Comment by tillrohrmann** [comment]
@uce I think you can close this PR. Thanks.

**Pull request closed without merging**